File: typescript/packages/core/src/commands/builtin/langfuse/grep.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import type { LangfuseAccessor } from '../../../accessor/langfuse.ts'
import type { IndexCacheStore } from '../../../cache/index/store.ts'
import {
  fetchDatasets,
  fetchPrompts,
  fetchSessions,
  fetchTraces,
} from '../../../core/langfuse/_client.ts'
import { resolveLangfuseGlob } from '../../../core/langfuse/glob.ts'
import { read as langfuseRead } from '../../../core/langfuse/read.ts'
import { readdir as langfuseReaddir } from '../../../core/langfuse/readdir.ts'
import { detectScope } from '../../../core/langfuse/scope.ts'
import { stat as langfuseStat } from '../../../core/langfuse/stat.ts'
import { IOResult } from '../../../io/types.ts'
import { type FileStat, type PathSpec, ResourceName } from '../../../types.ts'
import { command, type CommandFnResult, type CommandOpts } from '../../config.ts'
import { specOf } from '../../spec/builtins.ts'
import { grepGeneric } from '../generic/grep.ts'
import { compilePattern, patternArg } from '../grep_helper.ts'
import { formatRecords } from '../utils/output.ts'
import { fileReadProvision } from './_provision.ts'

function pickString(record: Record<string, unknown>, key: string): string {
  const value = record[key]
  return typeof value === 'string' ? value : ''
}

function filterTraces(
  traces: readonly Record<string, unknown>[],
  pattern: RegExp,
): CommandFnResult {
  const lines: string[] = []
  for (const t of traces) {
    const traceId = pickString(t, 'id')
    const lineJson = JSON.stringify(t)
    if (!pattern.test(lineJson)) continue
    lines.push(`traces/${traceId}.json:${lineJson}`)
  }
  if (lines.length === 0) return [new Uint8Array(0), new IOResult({ exitCode: 1 })]
  return [formatRecords(lines), new IOResult()]
}

function filterSessions(
  sessions: readonly Record<string, unknown>[],
  pattern: RegExp,
): CommandFnResult {
  const lines: string[] = []
  for (const s of sessions) {
    const sessionId = pickString(s, 'id')
    if (!pattern.test(sessionId)) continue
    const lineJson = JSON.stringify(s)
    lines.push(`sessions/${sessionId}:${lineJson}`)
  }
  if (lines.length === 0) return [new Uint8Array(0), new IOResult({ exitCode: 1 })]
  return [formatRecords(lines), new IOResult()]
}

function filterPrompts(
  prompts: readonly Record<string, unknown>[],
  pattern: RegExp,
): CommandFnResult {
  const lines: string[] = []
  const seen = new Set<string>()
  for (const p of prompts) {
    const promptName = pickString(p, 'name')
    if (seen.has(promptName)) continue
    if (!pattern.test(promptName)) continue
    seen.add(promptName)
    const lineJson = JSON.stringify(p)
    lines.push(`prompts/${promptName}:${lineJson}`)
  }
  if (lines.length === 0) return [new Uint8Array(0), new IOResult({ exitCode: 1 })]
  return [formatRecords(lines), new IOResult()]
}

function filterDatasets(
  datasets: readonly Record<string, unknown>[],
  pattern: RegExp,
): CommandFnResult {
  const lines: string[] = []
  for (const d of datasets) {
    const datasetName = pickString(d, 'name')
    if (!pattern.test(datasetName)) continue
    const lineJson = JSON.stringify(d)
    lines.push(`datasets/${datasetName}:${lineJson}`)
  }
  if (lines.length === 0) return [new Uint8Array(0), new IOResult({ exitCode: 1 })]
  return [formatRecords(lines), new IOResult()]
}

async function* langfuseStream(
  accessor: LangfuseAccessor,
  p: PathSpec,
  index?: IndexCacheStore,
): AsyncIterable<Uint8Array> {
  yield await langfuseRead(accessor, p, index)
}

async function grepCommand(
  accessor: LangfuseAccessor,
  paths: PathSpec[],
  texts: string[],
  opts: CommandOpts,
): Promise<CommandFnResult> {
  const pattern = patternArg(texts, opts.flags)
  const limit = accessor.config.defaultSearchLimit ?? 50

  const first = paths[0]
  if (first !== undefined && pattern !== null) {
    const scope = detectScope(first)
    const ignoreCase = opts.flags.i === true
    const fixedString = opts.flags.F === true
    const wholeWord = opts.flags.w === true
    const pat = compilePattern(pattern, ignoreCase, fixedString, wholeWord)
    if (scope.level === 'traces' || scope.level === 'root') {
      const traces = await fetchTraces(accessor.transport, { limit })
      return filterTraces(traces, pat)
    }
    if (scope.level === 'sessions') {
      const sessions = await fetchSessions(accessor.transport, { limit })
      return filterSessions(sessions, pat)
    }
    if (scope.level === 'prompts') {
      const prompts = await fetchPrompts(accessor.transport)
      return filterPrompts(prompts, pat)
    }
    if (scope.level === 'datasets') {
      const datasets = await fetchDatasets(accessor.transport)
      return filterDatasets(datasets, pat)
    }
  }

  const resolved =
    paths.length > 0 ? await resolveLangfuseGlob(accessor, paths, opts.index ?? undefined) : []
  const stat = (p: PathSpec): Promise<FileStat> =>
    langfuseStat(accessor, p, opts.index ?? undefined)
  const readdir = (p: PathSpec): Promise<string[]> =>
    langfuseReaddir(accessor, p, opts.index ?? undefined)
  return grepGeneric('grep', resolved, texts, opts, stat, readdir, (p) =>
    langfuseStream(accessor, p, opts.index ?? undefined),
  )
}

export const LANGFUSE_GREP = command({
  name: 'grep',
  resource: ResourceName.LANGFUSE,
  spec: specOf('grep'),
  fn: grepCommand,
  provision: fileReadProvision,
})

File: python/mirage/commands/builtin/langfuse/grep.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import json
import re
from collections.abc import AsyncIterator

from mirage.accessor.langfuse import LangfuseAccessor
from mirage.cache.index import IndexCacheStore
from mirage.commands.builtin.generic.grep import grep as generic_grep
from mirage.commands.builtin.grep_helper import compile_pattern, pattern_arg
from mirage.commands.builtin.langfuse._provision import file_read_provision
from mirage.commands.builtin.utils.output import format_records
from mirage.commands.registry import command
from mirage.commands.spec import SPECS
from mirage.commands.spec.types import FlagView
from mirage.core.langfuse._client import (fetch_datasets, fetch_prompts,
                                          fetch_sessions, fetch_traces)
from mirage.core.langfuse.glob import resolve_glob
from mirage.core.langfuse.read import read as langfuse_read
from mirage.core.langfuse.readdir import readdir as _readdir
from mirage.core.langfuse.scope import detect_scope
from mirage.core.langfuse.stat import stat as _stat
from mirage.io.types import ByteSource, IOResult
from mirage.provision.types import ProvisionResult
from mirage.types import PathSpec


def _filter_traces(
    traces: list[dict],
    pattern: re.Pattern,
) -> tuple[bytes, IOResult]:
    lines: list[str] = []
    for t in traces:
        trace_id = t.get("id", "")
        line_json = json.dumps(t, ensure_ascii=False)
        if not pattern.search(line_json):
            continue
        line = f"traces/{trace_id}.json:{line_json}"
        lines.append(line)
    if not lines:
        return b"", IOResult(exit_code=1)
    return format_records(lines), IOResult()


def _format_session_results(
    sessions: list[dict],
    pattern: re.Pattern,
) -> tuple[bytes, IOResult]:
    lines: list[str] = []
    for s in sessions:
        session_id = s.get("id", "")
        if not pattern.search(session_id):
            continue
        line_json = json.dumps(s, ensure_ascii=False)
        line = f"sessions/{session_id}:{line_json}"
        lines.append(line)
    if not lines:
        return b"", IOResult(exit_code=1)
    return format_records(lines), IOResult()


def _format_prompt_results(
    prompts: list[dict],
    pattern: re.Pattern,
) -> tuple[bytes, IOResult]:
    lines: list[str] = []
    seen: set[str] = set()
    for p in prompts:
        prompt_name = p.get("name", "")
        if prompt_name in seen:
            continue
        if not pattern.search(prompt_name):
            continue
        seen.add(prompt_name)
        line_json = json.dumps(p, ensure_ascii=False)
        line = f"prompts/{prompt_name}:{line_json}"
        lines.append(line)
    if not lines:
        return b"", IOResult(exit_code=1)
    return format_records(lines), IOResult()


def _format_dataset_results(
    datasets: list[dict],
    pattern: re.Pattern,
) -> tuple[bytes, IOResult]:
    lines: list[str] = []
    for d in datasets:
        dataset_name = d.get("name", "")
        if not pattern.search(dataset_name):
            continue
        line_json = json.dumps(d, ensure_ascii=False)
        line = f"datasets/{dataset_name}:{line_json}"
        lines.append(line)
    if not lines:
        return b"", IOResult(exit_code=1)
    return format_records(lines), IOResult()


async def grep_provision(
    accessor: LangfuseAccessor,
    paths: list[PathSpec],
    *texts: str,
    **_extra: object,
) -> ProvisionResult:
    return await file_read_provision(
        accessor, paths,
        "grep " + " ".join(texts + tuple(str(p) for p in paths)))


@command("grep",
         resource="langfuse",
         spec=SPECS["grep"],
         provision=grep_provision)
async def grep(
    accessor: LangfuseAccessor,
    paths: list[PathSpec],
    *texts: str,
    stdin: AsyncIterator[bytes] | bytes | None = None,
    prefix: str = "",
    index: IndexCacheStore = None,
    **flags: object,
) -> tuple[ByteSource | None, IOResult]:
    fl = FlagView(flags, spec=SPECS["grep"])
    pattern = pattern_arg(texts, fl)

    limit = accessor.config.default_search_limit

    if paths and pattern is not None:
        scope = detect_scope(paths[0])
        ignore_case = fl.bool("i")
        fixed_string = fl.bool("F")
        whole_word = fl.bool("w")

        if scope.level == "traces" or scope.level == "root":
            traces = await fetch_traces(
                accessor.api,
                limit=limit,
            )
            pat = compile_pattern(pattern, ignore_case, fixed_string,
                                  whole_word)
            return _filter_traces(traces, pat)

        if scope.level == "sessions":
            sessions = await fetch_sessions(
                accessor.api,
                limit=limit,
            )
            pat = compile_pattern(pattern, ignore_case, fixed_string,
                                  whole_word)
            return _format_session_results(sessions, pat)

        if scope.level == "prompts":
            prompts = await fetch_prompts(accessor.api)
            pat = compile_pattern(pattern, ignore_case, fixed_string,
                                  whole_word)
            return _format_prompt_results(prompts, pat)

        if scope.level == "datasets":
            datasets = await fetch_datasets(accessor.api)
            pat = compile_pattern(pattern, ignore_case, fixed_string,
                                  whole_word)
            return _format_dataset_results(datasets, pat)

    resolved = await resolve_glob(accessor, paths,
                                  index=index) if paths else []
    return await generic_grep(
        resolved,
        texts,
        flags,
        readdir=_readdir,
        stat=_stat,
        read_bytes=langfuse_read,
        read_stream=None,
        accessor=accessor,
        stdin=stdin,
        index=index,
    )
